packages/jupyter-ai/jupyter_ai/handlers.py

```diff
@@ -571,3 +575,115 @@ def get(self):
         # sort slash commands by slash id and deliver the response
         response.slash_commands.sort(key=lambda sc: sc.slash_id)
         self.finish(response.json())
+
+
+class AutocompleteOptionsHandler(BaseAPIHandler):
+    """List context that are currently available to the user."""
+
+    @property
+    def config_manager(self) -> ConfigManager:  # type:ignore[override]
+        return self.settings["jai_config_manager"]
+
+    @property
+    def context_providers(self) -> Dict[str, "BaseContextProvider"]:
+        return self.settings["jai_context_providers"]
+
+    @property
+    def chat_handlers(self) -> Dict[str, "BaseChatHandler"]:
+        return self.settings["jai_chat_handlers"]
+
+    @web.authenticated
+    def get(self):
+        response = ListOptionsResponse()
+
+        # if no selected LLM, return an empty response
+        if not self.config_manager.lm_provider:
+            self.finish(response.json())
+            return
+
+        response.options = (
+            self._get_slash_command_options() + self._get_context_provider_options()
+        )
+        self.finish(response.json())
+
+    @web.authenticated
+    def post(self):
```

Comment:
The HTTP API for autocomplete options can be simplified and be made more RESTful. Since getting command options & command arguments both involve reading data, they should both use the `GET` HTTP verb. `POST` should be reserved for creating a data entity.

Can you refactor the GET handler to accept a URL query argument, like `?partialCommand=...`? [This should be accessible via `self.get_query_argument()` here.](https://www.tornadoweb.org/en/stable/web.html#tornado.web.RequestHandler.get_query_argument) When this query argument is specified, this handler should return a list of valid commands that share the prefix `partialCommand`, thus allowing both capabilities to be implemented under `GET chats/autocomplete_options`.

Reply:
Ah thanks a lot for this. I knew what I originally was doing was incorrect but I wasn't sure how to integrate the dual logic. Using POST was just a way to quickly implement it without having to create another handler. I just totally forgot to call it out.